request.go

```diff
@@ -980,7 +994,20 @@ func (r *Request) sendRequest() *http.Response {
 		return nil
 	}
 
-	resp, err := r.config.Client.Do(r.http)
+	var err error
+	var resp *http.Response
+	if r.maxRetry == 0 {
+		resp, err = r.config.Client.Do(r.http)
+	} else {
+		for i := 0; i < r.maxRetry; i++ {
+			resp, err = r.config.Client.Do(r.http)
```

Comment:
The problem here is that, in general, http.Request is not reusable after calling http.Client.Do, even if it failed. In particular, http.Request.Body is io.Reader, and http.Client may read the whole body, so that the retry will see an EOF, like the body is empty.

This problem doesn't appear with requests without body. We should add a test with a POST request with non-empty body which will catch this bug.

The typical way to solve this problem is:

1. read the whole body into buffer
2. create a new reader from the buffer
3. call Client.Do
4. on failure, go to step (2)

Some optimizations are possible, e.g. if the client buffer is ReaderSeeker, we can just seek to the beginning before each retry.

Quick googling showed that there is also a package the handles this issue: https://github.com/hashicorp/go-retryablehttp

Maybe we can use it here? Or at least reproduce part of its logic.